### DISCUSSION TITLE
Feature Request: support bitmap datatype and functions

### DISCUSSION THREAD

**Discussion by vkingnew** [duckdb/duckdb#13395] [Ideas]
https://github.com/RoaringBitmap/RoaringBitmap
this function had implemented in clickhouse and apache doris and so on

**Comment by rpbouman** [comment]
Not sure I understand - how is that different from the `BITSTRING` data type (https://duckdb.org/docs/sql/data_types/bitstring) and functions (https://duckdb.org/docs/sql/functions/bitstring.html)?

**Reply by coracuity** [reply]
A roaring bitmap is a *compressed* bitmap, which is particularly helpful in the case of sparse bitmaps. Bitstrings would be problematic for sparse bitmaps in the u32 range that roaring bitmaps support (bitstring would be something like 512 MiB, regardless of how few bits are flipped on 😬)

**Reply by Tishj** [reply]
We can likely support compressing it with Roaring (compression used for validity masks)

But I don't see why an extra type would be required. As mentioned, the BITSTRING type already represents a bitset

**Reply by Tishj** [reply]
We were already thinking of using this compression for BOOL columns. If that's done then I imagine ARRAY[BOOL] could be a more performant way to represent the bitmap, assuming they're of a fixed size (?)

**Reply by vkingnew** [reply]
bitstring is alias of bit, looks that can store 0s and 1s. But accord to the [RoaringBitmap] (https://github.com/RoaringBitmap/RoaringBitmap) it can store int  0 ~ 18446744073709551615. THE related bitmap function can refer to:
https://docs.starrocks.io/docs/category/bitmap/
https://clickhouse.com/docs/sql-reference/functions/bitmap-functions

**Reply by coracuity** [reply]
the purpose of roaring bitmaps for me would be to precalculate *part* of a count distinct and store it as a value. this would be a count distinct where values are in the u32 range, so it fits my problem nicely. later I'd combine it with other values in the same set by way of aggregation. it can dramatically reduce the number of records you need to process later when you need to query and aggregate

**Reply by Tishj** [reply]
@vkingnew bitstring is binary, roaring bitmaps are binary
The only difference is how this is internally stored, which I've mentioned can be solved by using a better compression algorithm (we have implemented Roaring at the compression layer, though it's not used for BITSTRING yet)

`18446744073709551615` in binary is represented as `1111111111111111111111111111111111111111111111111111111111111111` which is only 64 bits, which means a bitstring of size 64 could represent this just fine.

You could build an extension to add the BITMAP type, using a BITSTRING as underlying type, similar to what the JSON extension does (VARCHAR type with the "json" alias)
